client/test/client.go

```diff
@@ -112,7 +133,90 @@ func (c *Client) Send(recipient common.Address, amount *big.Int) error {
 	})
 	log.Debug("Sending TX")
 	defer log.Trace("TX sent")
-	return c.tr.Send(tx)
+	err := c.tr.Send(tx)
+	if err == nil {
+		c.balance.Sub(c.balance, amount)
+	}
+	return err
+}
+
+// SendToClient sends amount to the given recipient.
+// Use this if you need proper balance tracking in your tests.
+func (c *Client) SendToClient(recipient *Client, amount *big.Int) error {
+	err := c.Send(recipient.Address(), amount)
+	if err == nil {
+		recipient.balance.Add(recipient.balance, amount)
+	}
+	return err
+}
+
+// SendInvalidTxs sends several invalid transactions. It uses InvalidTxs, see
+// its documentation for which invalid transactions are used.
+func (c *Client) SendInvalidTxs(rng *rand.Rand, validRecipient common.Address) (errs []error) {
+	for _, tx := range c.InvalidTxs(rng, validRecipient) {
+		errs = append(errs, c.tr.Send(tx))
+	}
+	return errs
+}
+
+// InvalidTxs generates a list of several invalid transactions, each having one
+// of the following fields set to an invalid value:
+// - Sig (random)
+// - Nonce (too high)
+// - Epoch (future)
+// - Recipient (random)
+// - Sender (random)
+// - Amount (too much)
+// The remaining fields are set to a valid value.
+func (c *Client) InvalidTxs(rng *rand.Rand, validRecipient common.Address) (txs []*tee.Transaction) {
+	valid := tee.Transaction{
+		Nonce:     c.Nonce + 1, // only increment locally
+		Epoch:     c.params.TxEpoch(c.minedBlockNum + 1),
+		Sender:    c.ethClient.Account().Address,
+		Recipient: validRecipient,
+		Amount:    c.Balance(), // total balance is valid
+	}
+
+	// 0. Invalid signature
+	tx0 := valid
+	tx0.Sig = make([]byte, 65)
+	rng.Read(tx0.Sig)
+	txs = append(txs, &tx0)
+
+	signAppend := func(tx *tee.Transaction) {
+		c.SignTx(tx)
+		txs = append(txs, tx)
+	}
+
+	// 1. Invalid nonce
+	tx1 := valid
+	tx1.Nonce += 42
```

Comment:
I have the feeling it might be better to test the edge cases, where the `Nonce` (and everywhere else, where applicable) is exactly `1` more and/or less than expected.

Since off-by-one errors are more common and this would also catch places where an implementer mixed up `{<|>|>=|<=}`.

Reply:
good point, will do that

Reply:
This inspired me to also check for negative amounts and we didn't check that yet :joy: fixed now.